### ISSUE TITLE
Requisite races with dependent unit when after/before ordering is not present

### ISSUE THREAD

**Issue by msoltyspl** [systemd/systemd#8309]
- bug report
- systemd 237.64
- archlinux

I'm not 100% if this is considered a bug (and obviously it's 100% correct for 'requires'), but no matter how I look at it, it feels wrong for 'requisite'. Consider 2 trivial services:
```
[Unit]
Description=A

[Service]
Type=simple
ExecStart=/bin/echo A
ExecStop=/bin/true
```
and 
```
[Unit]
Description=B
Requisite=a.service

[Service]
Type=simple
ExecStart=/bin/echo B
ExecStop=/bin/true
```

Add those, do daemon-reload - then do systemctl start b.service. It will sometimes start, sometimes bail out:
```
-- Logs begin at Tue 2016-12-06 11:31:40 CET, end at Wed 2018-02-28 15:40:12 CET. --
Feb 28 15:35:00 hakai systemd[1]: Started B.
Feb 28 15:35:00 hakai echo[22204]: B
Feb 28 15:35:03 hakai systemd[1]: Dependency failed for B.
Feb 28 15:35:03 hakai systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
Feb 28 15:35:09 hakai systemd[1]: Started B.
Feb 28 15:35:09 hakai echo[22215]: B
Feb 28 15:35:13 hakai systemd[1]: Started B.
Feb 28 15:35:13 hakai echo[22221]: B
Feb 28 15:35:14 hakai systemd[1]: Dependency failed for B.
Feb 28 15:35:14 hakai systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
Feb 28 15:35:14 hakai systemd[1]: Started B.
Feb 28 15:35:14 hakai echo[22231]: B
Feb 28 15:35:14 hakai systemd[1]: Dependency failed for B.
Feb 28 15:35:14 hakai systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
Feb 28 15:35:15 hakai systemd[1]: Started B.
Feb 28 15:35:15 hakai echo[22242]: B
Feb 28 15:35:15 hakai systemd[1]: Dependency failed for B.
Feb 28 15:35:15 hakai systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
Feb 28 15:35:15 hakai systemd[1]: Started B.
Feb 28 15:35:15 hakai echo[22254]: B
Feb 28 15:35:16 hakai systemd[1]: Started B.
Feb 28 15:35:16 hakai echo[22260]: B
Feb 28 15:35:16 hakai systemd[1]: Started B.
Feb 28 15:35:16 hakai echo[22266]: B
Feb 28 15:35:16 hakai systemd[1]: Started B.
Feb 28 15:35:16 hakai echo[22272]: B
Feb 28 15:35:17 hakai systemd[1]: Dependency failed for B.
Feb 28 15:35:17 hakai systemd[1]: b.service: Job b.service/start failed with result 'dependency'.
```
and relevant log for A:
```
Feb 28 15:35:00 hakai systemd[1]: A is not active.
Feb 28 15:35:03 hakai systemd[1]: A is not active.
Feb 28 15:35:09 hakai systemd[1]: A is not active.
Feb 28 15:35:13 hakai systemd[1]: A is not active.
Feb 28 15:35:14 hakai systemd[1]: A is not active.
Feb 28 15:35:14 hakai systemd[1]: A is not active.
Feb 28 15:35:14 hakai systemd[1]: A is not active.
Feb 28 15:35:15 hakai systemd[1]: A is not active.
Feb 28 15:35:15 hakai systemd[1]: A is not active.
Feb 28 15:35:15 hakai systemd[1]: A is not active.
Feb 28 15:35:16 hakai systemd[1]: A is not active.
Feb 28 15:35:16 hakai systemd[1]: A is not active.
Feb 28 15:35:16 hakai systemd[1]: A is not active.
Feb 28 15:35:17 hakai systemd[1]: A is not active.
```

This essentially looks like if requisite was treated as requires initially - and without explicit before/after - if "start b" even occurs before "start a" would have (no ordering), then it considers it success for requisite as well.

If we don't cosider it a bug, then I would submit a small man page update clarifying that this kind of behaviour can be present for 'requisite' in case of missing before/after ordering.

**Comment by boucman** [comment]
Requires now imply After, but IIRC this is new since 237.

I'm not a dev but I think it would make sense for Requisite to imply After for the same reason... you always want those two together anywah.

**Comment by arvidjaar** [comment]
@msoltyspl 

Same is true for `Requires`. It is not a bug per se (if "bug" means "does not work as intended"), but certainly design that turned out to be very confusing for users once systemd got wider acceptance.

@boucman 

> Requires now imply After

Reference please? I do not see anything to this effect in code, commit log or NEWS. It has always been true for **target** with `DefaultDependencies=true`, may be that is what you mean?

**Comment by boucman** [comment]
I was pretty convinced I saw taht in the last release, but I probably misremembered, sorry for the noise

**Comment by poettering** [comment]
We can certainly make this clearer in the docs: ordering and requirements are orthogonal in systemd. That's true for Requires= and also for Requisite=. In the docs we don't repeat that fact for Requisite= but I guess we could

**Comment by poettering** [comment]
Fix waiting in #9060 btw.
